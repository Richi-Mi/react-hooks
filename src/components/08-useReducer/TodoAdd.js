import React from 'react';
import { Fragment } from 'react/cjs/react.development';
import PropTypes from 'prop-types';
import useForm from '../../hooks/useForm';

const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset ] = useForm({
        description: ''
    });

    const handleSubmit = ( ev, descrip = description ) => {
        ev.preventDefault();

        if (descrip.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: descrip,
            done: false
        }
        
        reset();
        handleAddTodo( newTodo );
    }

    return (
        <Fragment>
            <h4> Agregar Tarea </h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name='description'
                    placeholder='Aprender...'
                    autoComplete='off'
                    className='form-control mb-3'
                    onChange={ ( ev ) => handleInputChange( ev ) }
                    value={ description }
                />
                <button
                    className='btn btn-outline-primary form-control'
                    type='submit'
                >
                    Agregar
                </button>
            </form>
        </Fragment>
    )
}

TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
}

export default TodoAdd
