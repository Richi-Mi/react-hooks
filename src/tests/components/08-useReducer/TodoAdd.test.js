import { shallow } from 'enzyme';
import React from 'react';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe(' * Pruebas sobre el Componente <TodoAdd /> ', () => {

    const handleAddTodo = jest.fn();
    const wraper = shallow( <TodoAdd handleAddTodo = { handleAddTodo } /> );

    test(' - Debe de mostrar el componente correctamente', () => {

        expect( wraper ).toMatchSnapshot(); 
    
    })
    test(' - No debe de llamar handleAddTodo() si no hay argumentos', () => {
    
        const formSubmit = wraper.find('form').prop('onSubmit');
        formSubmit({ preventDefault: () => {} });

        // expect( handleAddTodo ).not.toHaveBeenCalled();
        expect( handleAddTodo ).toHaveBeenCalledTimes(0);
    
    }) 
    test(' - Debe de llamar handleAddTodo() con argumentos', () => {
    
        const value = 'Aprender MongoDB';
        const formSubmit = wraper.find('form').prop('onSubmit');
        wraper.find('input').simulate('change', {
            target: {
                name: 'description',
                value
            }
        });
        console.log( wraper.find('input').prop('value') )
        formSubmit({ preventDefault: () => {} }, value);

        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any( Object ) );
        expect( handleAddTodo ).toHaveBeenCalledWith({
            id: expect.any( Number ),
            desc: value,
            done: false
        });

        expect( wraper.find('input').prop('value') ).toBe('');
    });
});
