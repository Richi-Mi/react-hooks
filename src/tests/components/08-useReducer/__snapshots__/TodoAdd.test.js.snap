// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` * Pruebas sobre el Componente <TodoAdd />   - Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h4>
     Agregar Tarea 
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control mb-3"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary form-control"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
